sqlgen/db_test.go

```diff
@@ -0,0 +1,538 @@
+package sqlgen
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type limitTestcase struct {
+	title           string
+	withFilterLimit func(*DB, Filter) *DB
+}
+
+func getLimitTestcases(t *testing.T, operation string) []limitTestcase {
+	return []limitTestcase{
+		{
+			title: fmt.Sprintf("%s with shard limit", operation),
+			withFilterLimit: func(db *DB, filter Filter) *DB {
+				db, err := db.WithShardLimit(filter)
+				assert.NoError(t, err)
+				return db
+			},
+		},
+		{
+			title: fmt.Sprintf("%s with dynamic limit", operation),
+			withFilterLimit: func(db *DB, filter Filter) *DB {
+				dynamicLimitFunc := func(ctx context.Context, table string) Filter {
+					// We can be generic here, not testing dynamic limits specifically.
+					assert.Contains(t, []string{"users", "just_ids"}, table)
+					return filter
+				}
+				dynamicLimitErrorFunc := func(err error) error {
+					return err
+				}
+				db, _ = db.WithDynamicLimit(
+					dynamicLimitFunc,
+					dynamicLimitErrorFunc,
+				)
+				return db
+			},
+		},
+	}
+}
+
+func TestShardLimitTwice(t *testing.T) {
+	tdb, db, err := setup()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tdb.Close()
+	aliceDb, err := db.WithShardLimit(Filter{
+		"name": "Alice",
+	})
+	if err != nil {
+		t.Error(err)
+	}
+
+	if _, err := aliceDb.WithShardLimit(nil); err == nil || err.Error() != "already has shard limit" {
+		t.Error("could double shard limit")
+	}
+}
+
+func TestDynamicLimitTwice(t *testing.T) {
+	tdb, db, err := setup()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dynamicLimitFunc := func(ctx context.Context, table string) Filter {
+		return Filter{
+			"name": "Alice",
+		}
+	}
+	dynamicLimitErrorFunc := func(err error) error {
+		return err
+	}
+
+	defer tdb.Close()
+	aliceDb, err := db.WithDynamicLimit(dynamicLimitFunc, dynamicLimitErrorFunc)
+	if err != nil {
```

Comment:
nit: `assert.NoError(t, err)`

Reply:
Updated the entire unit test to have better `assert` hygiene.